### PR TITLE
add new functions

### PR DESCRIPTION
**Pull request by wjwwood**
Connects to ros2/rmw#154

### CODE REVIEW

File: rmw_implementation/src/functions.cpp [hunk 1]
```diff
@@ -441,11 +461,19 @@ void prefetch_symbols(void)
...
+  if (!symbol_rmw_init) {
+    symbol_rmw_init = get_symbol("rmw_init");
+  }
+  if (!symbol_rmw_init) {
+    return RMW_RET_ERROR;
+  }
+
+  typedef rmw_ret_t (* FunctionSignature)(const rmw_init_options_t *, rmw_context_t *);
+  FunctionSignature func = reinterpret_cast<FunctionSignature>(symbol_rmw_init);
+  return func(options, context);
```
Comment (wjwwood) [inline rmw_implementation/src/functions.cpp]: @dirk-thomas do you know how to make this work with `CALL_SYMBOL`, I could figure out how to go from 0 args to >=1 args, so I expanded it manually.

Comment (dirk-thomas) [reply rmw_implementation/src/functions.cpp]: How did you try to invoke it with one argument? What was the error message?

Comment (wjwwood) [reply rmw_implementation/src/functions.cpp]: I don't have it on hand, it was a few days ago. Let me see if I can reproduce it.

Comment (wjwwood) [reply rmw_implementation/src/functions.cpp]: I tried something like this:

```c++
CALL_SYMBOL(
    rmw_init, rmw_ret_t, RMW_RET_ERROR,
    ARG_TYPES(const rmw_init_options_t *, rmw_context_t *), ARG_VALUES_2(options, context))
```

```
/Users/william/ros2_ws/src/ros2/rmw_implementation/rmw_implementation/src/functions.cpp:469:61: error: use of undeclared identifier 'v2'
    ARG_TYPES(const rmw_init_options_t *, rmw_context_t *), ARG_VALUES_2(options, context))
                                                            ^
/Users/william/ros2_ws/src/ros2/rmw_implementation/rmw_implementation/src/functions.cpp:182:31: note: expanded from macro 'ARG_VALUES_2'
#define ARG_VALUES_2(t2, ...) v2, EXPAND(ARG_VALUES_1(__VA_ARGS__))
                              ^
/Users/william/ros2_ws/src/ros2/rmw_implementation/rmw_implementation/src/functions.cpp:469:61: error: use of undeclared identifier 'v1'
/Users/william/ros2_ws/src/ros2/rmw_implementation/rmw_implementation/src/functions.cpp:182:42: note: expanded from macro 'ARG_VALUES_2'
#define ARG_VALUES_2(t2, ...) v2, EXPAND(ARG_VALUES_1(__VA_ARGS__))
                                         ^
/Users/william/ros2_ws/src/ros2/rmw_implementation/rmw_implementation/src/functions.cpp:181:26: note: expanded from macro 'ARG_VALUES_1'
#define ARG_VALUES_1(t1) v1
                         ^
/Users/william/ros2_ws/src/ros2/rmw_implementation/rmw_implementation/src/functions.cpp:464:37: warning: unused parameter 'options' [-Wunused-parameter]
rmw_init(const rmw_init_options_t * options, rmw_context_t * context)
                                    ^
/Users/william/ros2_ws/src/ros2/rmw_implementation/rmw_implementation/src/functions.cpp:464:62: warning: unused parameter 'context' [-Wunused-parameter]
rmw_init(const rmw_init_options_t * options, rmw_context_t * context)
                                                             ^
2 warnings and 2 errors generated.
```

And I tried:

```c++
CALL_SYMBOL(
    rmw_init, rmw_ret_t, RMW_RET_ERROR,
    ARG_TYPES(const rmw_init_options_t *, rmw_context_t *), EXPAND(ARG_VALUES_2(options, context)))
```

```
/Users/william/ros2_ws/src/ros2/rmw_implementation/rmw_implementation/src/functions.cpp:469:68: error: use of undeclared identifier 'v2'
    ARG_TYPES(const rmw_init_options_t *, rmw_context_t *), EXPAND(ARG_VALUES_2(options, context)))
                                                                   ^
/Users/william/ros2_ws/src/ros2/rmw_implementation/rmw_implementation/src/functions.cpp:182:31: note: expanded from macro 'ARG_VALUES_2'
#define ARG_VALUES_2(t2, ...) v2, EXPAND(ARG_VALUES_1(__VA_ARGS__))
                              ^
/Users/william/ros2_ws/src/ros2/rmw_implementation/rmw_implementation/src/functions.cpp:469:68: error: use of undeclared identifier 'v1'
/Users/william/ros2_ws/src/ros2/rmw_implementation/rmw_implementation/src/functions.cpp:182:42: note: expanded from macro 'ARG_VALUES_2'
#define ARG_VALUES_2(t2, ...) v2, EXPAND(ARG_VALUES_1(__VA_ARGS__))
                                         ^
/Users/william/ros2_ws/src/ros2/rmw_implementation/rmw_implementation/src/functions.cpp:181:26: note: expanded from macro 'ARG_VALUES_1'
#define ARG_VALUES_1(t1) v1
                         ^
/Users/william/ros2_ws/src/ros2/rmw_implementation/rmw_implementation/src/functions.cpp:464:37: warning: unused parameter 'options' [-Wunused-parameter]
rmw_init(const rmw_init_options_t * options, rmw_context_t * context)
                                    ^
/Users/william/ros2_ws/src/ros2/rmw_implementation/rmw_implementation/src/functions.cpp:464:62: warning: unused parameter 'context' [-Wunused-parameter]
rmw_init(const rmw_init_options_t * options, rmw_context_t * context)
                                                             ^
2 warnings and 2 errors generated.
```

So I thought maybe the arguments had to be named just so, so I tried:

```
rmw_init(const rmw_init_options_t * v1, rmw_context_t * v2)
{
  prefetch_symbols();
  CALL_SYMBOL(
    rmw_init, rmw_ret_t, RMW_RET_ERROR,
    ARG_TYPES(const rmw_init_options_t *, rmw_context_t *), ARG_VALUES_2(v1, v2))
```

```
/Users/william/ros2_ws/src/ros2/rmw_implementation/rmw_implementation/src/functions.cpp:469:61: error: cannot initialize a parameter of type 'const rmw_init_options_t *' with an lvalue of type 'rmw_context_t *'
    ARG_TYPES(const rmw_init_options_t *, rmw_context_t *), ARG_VALUES_2(v1, v2))
                                                            ^~~~~~~~~~~~~~~~~~~~
/Users/william/ros2_ws/src/ros2/rmw_implementation/rmw_implementation/src/functions.cpp:182:31: note: expanded from macro 'ARG_VALUES_2'
#define ARG_VALUES_2(t2, ...) v2, EXPAND(ARG_VALUES_1(__VA_ARGS__))
                              ^~
/Users/william/ros2_ws/src/ros2/rmw_implementation/rmw_implementation/src/functions.cpp:207:15: note: expanded from macro 'CALL_SYMBOL'
  return func(arg_values);
              ^~~~~~~~~~
1 error generated.
```

Which sort of works but then the arguments are backwards... I tried to understand why that was happening, but I couldn't figure out what I was supposed to do or why they were reversed (or if they actually needed to be named `v1` and `v2`).

**Pull request merged**